src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,276 @@
+        "",
+        (secure ? (maybe_secure_port ? *maybe_secure_port : DBMS_DEFAULT_SECURE_PORT) : context->getTCPPort()),
+        false /* treat_local_as_remote */,
+        context->getApplicationType() == Context::ApplicationType::LOCAL /* treat_local_port_as_remote */,
+        secure);
+    return cluster;
+}
+
+/// Reads data from zookeeper and tries to update cluster.
+/// Returns true on success (or no update required).
+bool ClusterDiscovery::updateCluster(ClusterInfo & cluster_info)
+{
+    LOG_TRACE(log, "Updating cluster '{}'", cluster_info.name);
+
+    auto zk = context->getZooKeeper();
+
+    int start_version;
+    Strings node_uuids = getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &start_version, false);
+
+    if (node_uuids.empty())
+    {
+        LOG_ERROR(log, "Can't find any node in cluster '{}', will register again", cluster_info.name);
+        registerInZk(zk, cluster_info);
+        return false;
+    }
+
+    auto & nodes_info = cluster_info.nodes_info;
+    if (!needUpdate(node_uuids, nodes_info))
+    {
+        LOG_TRACE(log, "No update required for cluster '{}'", cluster_info.name);
+        return true;
+    }
+
+    nodes_info = getNodes(zk, cluster_info.zk_root, node_uuids);
+
+    if (nodes_info.empty())
+        return false;
+
+    int current_version;
+    getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &current_version, true);
+
+    if (current_version != start_version)
+    {
+        nodes_info.clear();
+        return false;
+    }
+
+    auto cluster = getCluster(cluster_info);
+    context->setCluster(cluster_info.name, cluster);
+    return true;
+}
+
+bool ClusterDiscovery::updateCluster(const String & cluster_name)
+{
+    auto cluster_info = clusters_info.find(cluster_name);
+    if (cluster_info == clusters_info.end())
+    {
+        LOG_ERROR(log, "Unknown cluster '{}'", cluster_name);
+        return false;
+    }
+    return updateCluster(cluster_info->second);
+}
+
+void ClusterDiscovery::registerInZk(zkutil::ZooKeeperPtr & zk, ClusterInfo & info)
+{
+    String node_path = getReplicasListPath(info.zk_root) / node_name;
+    zk->createAncestors(node_path);
+
+    String payload = getFQDNOrHostName() + ":" + toString(server_port);
+
+    zk->createOrUpdate(node_path, payload, zkutil::CreateMode::Ephemeral);
+    LOG_DEBUG(log, "Current node {} registered in cluster {}", node_name, info.name);
+}
+
+void ClusterDiscovery::start()
+{
+    auto zk = context->getZooKeeper();
+
+    LOG_TRACE(log, "Starting working thread");
+    main_thread = ThreadFromGlobalPool([this] { runMainThread(); });
+
+    for (auto & [_, info] : clusters_info)
+    {
+        registerInZk(zk, info);
+        updateCluster(info);
+
+        if (!updateCluster(info))
+            LOG_WARNING(log, "Error on updating cluster '{}'", info.name);
+    }
+}
+
+void ClusterDiscovery::runMainThread()
+{
+    LOG_TRACE(log, "Worker thread started");
+    // setThreadName("ClusterDiscovery");
```

Comment:
Why commented? Seems useful.

Reply:
It blocks for some reason when I tried to run, I will check again

Reply:
I guess this is because it is too long (15 chars max), and so setThreadName throws, and you don't catch the error.